feat(fluid scroller): buffer scroll thresholds

hufort opened the pull request:
## Why was this needed
RBDND's usage in Services’ new Matrix highlighted a less-than-desirable behavior: [autoscroll](https://github.com/atlassian/react-beautiful-dnd/blob/master/docs/guides/auto-scrolling.md) kicks in when a the center of a draggable crosses a [threshold of 25%](https://github.com/atlassian/react-beautiful-dnd/blob/1a380855d0d008c5d6ef8e34c7c8ffb96c66a881/src/state/auto-scroller/fluid-scroller/config.js#L6) of the scroll container. This is fine when a user grabs a draggable located in the middle 50% of the container, but the behavior breaks down when the draggable is grabbed inside of one of those thresholds, as the container will unconditionally start scrolling that direction, even if the user’s intent is to scroll the opposite direction.

## How I did it
From a high-level, we essentially want to determine if a draggable originates within a scroll threshold, if it does, we prevent scrolling in that direction until the user expresses intent by dragging in that direction.

The preexisting behavior is broadly in use and doesn't seem to be causing issues at large, so it seemed advisable not to haul off and executively rewrite it with this new set of opinions.

Instead, the new behavior is disabled by default, but can be opted into by way of a new `fluidScrollerOptions` object that can be passed to `DragDropContext`. This options object takes the following properties, all of which are optional (and more thoroughly documented in [`type.js`](src/types.js)):
- `bufferThresholds` - enables the new behavior
- `bufferMinScroll` - the distance required to cause a scroll to occur if the draggable originates inside a buffered threshold, defaults to `24` (`px`)
- `configOverride` - allows an author to fine tune how the fluid scroller feels
- `thruGetScroll` - a function that receives a number of scroll props, allowing an author an entry point to modify or provide their own scroll logic — can be used in conjunction with `bufferThresholds: true`

Provided with `bufferThresholds: true` the bulk of the new behavior is provided by the following functions:
- [`calcAxisScrollConditions`](src/state/auto-scroller/fluid-scroller/get-scroll/buffer-thresholds/calc-axis-scroll-conditions.js) - calculates where a draggable is in relation to the thresholds
- [`bufferThresholds`](src/state/auto-scroller/fluid-scroller/get-scroll/buffer-thresholds/index.js) - uses the above mentioned conditions to either stomp on or allow a scroll change along a given axis

Beyond that, much of the diff here is made up by hoisting values, reorganizing code, adding or updating type definitions, threading the necessary props through from context, and consuming optional `configOverride` values.

## Specific Asks
- I implemented some optional chaining (`?.`) in my changes here. According to [caniuse](https://caniuse.com/?search=optional%20chaining), it appears that this is admissible by Planning Center standards. Or, that's my understanding at least.
- I haven't written test for this implementation yet, but I've kicked this around pretty extensively in Storybook, and I've also passed `{ bufferThresholds: true }` to all preexisting specs and run them; all green. That said, it may be prudent to have someone else fire up Storybook and kick things around.

## Demos
I'm including demos and code snippets for a few examples usages. There are a _lot_ of different scenarios you can potentially achieve via `fluidScrollerOptions`, so these are just a few to demo that the wiring works and that overrides take precedence as expected.

_Blue transparency added to help visualize thresholds_
### Old behavior
#### Grabbing a draggable inside a threshold causes autoscroll
https://user-images.githubusercontent.com/37631493/194932716-bea380b4-324c-4c4b-b0f9-61130035825a.mp4

### New behavior
#### Buffer thresholds - drag to scroll requires explicit intent
```js
const fluidScrollerOptions = {
    bufferThresholds: true,
  };
```
https://user-images.githubusercontent.com/37631493/194931248-f3713bdf-fc86-4065-b6ac-60bb24f1eddc.mp4

#### Buffer, increase threshold sizes and max scroll speed
```js
const fluidScrollerOptions = {
    bufferThresholds: true,
    configOverride: {
      startFromPercentage: 0.45,
      maxPixelScroll: 56,
    },
  };
```
https://user-images.githubusercontent.com/37631493/194931754-a20b3f1e-5d10-4e38-9595-1079ff2c4255.mp4

#### Buffer, increase threshold sizes, supply modified scroll behavior
```js
const fluidScrollerOptions = {
    bufferThresholds: true,
    thruGetScroll: ({ center, scroll, thresholdsVertical, windowScrollOffset }) => {
      if (center.y < thresholdsVertical.startScrollingFrom + windowScrollOffset.y ) {
        // reduce the scroll speed to 20% when scrolling up
        return { x: scroll.x / 5, y: scroll.y / 5 };
      }
      // double the scroll speed when scrolling down
      return { x: scroll.x * 2, y: scroll.y * 2 };
    },
    configOverride: {
      startFromPercentage: 0.45,
      maxPixelScroll: 56,
    },
  };
```
https://user-images.githubusercontent.com/37631493/194933963-a7633b9c-30f2-4d9a-aa6a-67c8d10d9667.mp4

shawnmaust reviewed (approved):
Nicely done!! I won't lie, with all of the files that got updated, I had a hard time keeping everything straight in my head 😅. I'm glad you made this an opt-in feature, too. It makes it way easier to ship it, knowing that it's not going to change any default behavior.

geolessel reviewed (approved):
Nice!

hufort commented:
@geolessel pushed a couple new commits since your review:
- updated the MAJOR version in `package.json`
- updated the `CHANGELOG` accordingly
- fixed a spec that was failing due to backtick-quoted `DragDropContext` in the `CHANGELOG`

Related questions
1. I wanted to verify that manually updating version-related things here was the correct course of action.
2. Assuming the above is correct, should I expect to follow through with [Tagging, Releasing, and Publishing](https://www.notion.so/planningcenter/Package-handling-at-Planning-Center-aka-How-we-npm-b1c78de88e124f7ca41955e0b082f55f#8423589b8a0c4a45b8436462a791b502) or will the FESE team be handling that?
